File: backend/src/commands/mlx/mlx_start.rs
use crate::commands::mlx::mlx_server_types::{
    MlxServerState, MlxServerStatus, MlxStartResult, Running,
};
use crate::errors::AppError;
use crate::inference::mlx::mlx_supported;
use crate::inference::mlx::server::mlx_locate::locate;
use crate::inference::mlx::server::mlx_runtime::{build_spawn_args, find_available_port, spawn_server};
use crate::inference::mlx::server::mlx_stderr::{spawn_stderr_reader, Phase};
use std::collections::VecDeque;
use std::sync::{Arc, Mutex};

const PORT_BASE: u16 = 8082;

/// Launch `mlx_lm.server` for a local model directory (downloaded via
/// `install_mlx_model`). Does NOT block on readiness — loading weights takes a
/// moment; the frontend polls health + `mlx_server_status` instead. Guards
/// against a double-spawn (stops a different running model first), picks a free
/// port, and starts the stderr reader so phase/death are observable.
#[tauri::command]
pub async fn start_mlx_server(
    state: tauri::State<'_, MlxServerState>,
    model_path: String,
) -> Result<MlxStartResult, AppError> {
    if !mlx_supported() {
        return Ok(MlxStartResult::StartFailed {
            error: "MLX needs Apple Silicon (macOS arm64).".into(),
        });
    }
    if state.is_model(&model_path) {
        return Ok(MlxStartResult::AlreadyRunning);
    }
    state.kill_all_servers().map_err(AppError::Internal)?;
    let configured = std::env::var("QUANTAMIND_MLX_SERVER").ok();
    let Some(exe) = locate(configured.as_deref()) else {
        return Ok(MlxStartResult::NotFound);
    };
    let Some(port) = find_available_port(PORT_BASE) else {
        return Ok(MlxStartResult::NoFreePort);
    };
    let mut child = match spawn_server(&exe, &build_spawn_args(&model_path, port)) {
        Ok(c) => c,
        Err(error) => return Ok(MlxStartResult::StartFailed { error }),
    };
    let pid = child.id();
    let phase = Arc::new(Mutex::new(Phase::Starting));
    let tail = Arc::new(Mutex::new(VecDeque::new()));
    if let Some(err) = child.stderr.take() {
        spawn_stderr_reader(err, phase.clone(), tail.clone());
    }
    state.store(Running { child, model: model_path, phase, tail }, port);
    Ok(MlxStartResult::Started { pid, port })
}

#[tauri::command]
pub async fn stop_mlx_server(state: tauri::State<'_, MlxServerState>) -> Result<(), AppError> {
    state.kill_all_servers().map_err(AppError::Internal)
}

#[tauri::command]
pub fn mlx_server_status(state: tauri::State<'_, MlxServerState>) -> MlxServerStatus {
    state.status()
}

File: frontend/src/shared/ipc/models/mlx_start.ts
import { invoke } from "@tauri-apps/api/core";
import { z } from "zod";

export const MlxStartResultSchema = z.discriminatedUnion("status", [
  z.object({ status: z.literal("already_running") }),
  z.object({
    status: z.literal("started"),
    pid: z.number().int().nonnegative(),
    port: z.number().int().nonnegative(),
  }),
  z.object({ status: z.literal("not_found") }),
  z.object({ status: z.literal("no_free_port") }),
  z.object({ status: z.literal("start_failed"), error: z.string() }),
]);
export type MlxStartResult = z.infer<typeof MlxStartResultSchema>;

export const MlxPhaseSchema = z.enum(["downloading", "starting", "ready", "exited"]);
export const MlxServerStatusSchema = z.discriminatedUnion("state", [
  z.object({ state: z.literal("stopped") }),
  z.object({ state: z.literal("running"), phase: MlxPhaseSchema, model: z.string() }),
  z.object({ state: z.literal("exited"), code: z.number().int().nullable(), stderr_tail: z.string() }),
]);
export type MlxServerStatus = z.infer<typeof MlxServerStatusSchema>;

/// Launch the app-managed mlx_lm.server on a local model directory (downloaded
/// via install_mlx_model). Returns immediately — readiness is polled via
/// mlxServerStatus + health.
export async function startMlxServer(modelPath: string): Promise<MlxStartResult> {
  return MlxStartResultSchema.parse(await invoke("start_mlx_server", { modelPath }));
}

export async function stopMlxServer(): Promise<void> {
  await invoke("stop_mlx_server");
}

export async function mlxServerStatus(): Promise<MlxServerStatus> {
  return MlxServerStatusSchema.parse(await invoke("mlx_server_status"));
}
